### PR TITLE
accept parameters in authenticate()

### LINKED ISSUE #9: Improve authenticate() method interface
It would help if we can pass a `token` parameter to the `authenticate()` function and in turn to `requestTicketValidation()`, thus yielding more control to the user over how the ticket is fetched. 

Currently the ticket can only be passed via querystring.

### PR DESCRIPTION
**Pull request by samsonradu**
This PR

* [ ] Have new test covering the feature
* [x] Add an argument to  `CasInterface::authenticate(array $parameters = []);` 
* [x] Does not introduce BC break

Fixes #9.

### CODE REVIEW

File: src/AbstractCas.php [hunk 1]
```diff
@@ -152,10 +152,6 @@ public function requestTicketValidation(
         array $parameters = [],
         ?ResponseInterface $response = null
     ): ?ResponseInterface {
-        if (false === $this->supportAuthentication()) {
```
Comment (drupol) [inline src/AbstractCas.php:155]: This change is out of scope of this PR, maybe we should open another PR and discuss about it ?

File: src/AbstractCas.php [hunk 2]
```diff
@@ -165,6 +161,10 @@ public function requestTicketValidation(
 
         $parameters += ['ticket' => $ticket];
 
+        if ('' === $parameters['ticket']) {
```
Comment (drupol) [inline src/AbstractCas.php]: Why this here?

File: src/CasInterface.php [hunk 3]
```diff
@@ -16,10 +16,13 @@ interface CasInterface
     /**
      * Authenticate the request.
      *
+     * @param array[] $parameters
+     *   The parameters related to the service. Can also contain a preset 'token'.
```
Comment (drupol) [inline src/CasInterface.php]: Token is not a valid parameter key for the CAS protocol, I would then remove that last sentence in order to not confuse any people.

**Review by drupol — CHANGES_REQUESTED** [review]
Hi @samsonradu ,

Thanks for the PR !

I did a quick review and left some comments.

Let me know what you think.

File: src/AbstractCas.php (hunk 2 above)
Comment (samsonradu) [reply src/AbstractCas.php]: This acts as a replacement for the `supportAuthentication()` which was removed above. Saves a round trip to ECAS servers with an empty token.

File: src/AbstractCas.php (hunk 1 above)
Comment (samsonradu) [reply src/AbstractCas.php:155]: Sure, we can do that. The reasoning for this removal is to be able to send tickets Authorization headers instead of query strings, but I understand we are bound by the CAS specs, though I believe the limitation is purely theoretical. There's nothing less safe than the querystring no? 

FYI
https://developer.github.com/changes/2019-11-05-deprecated-passwords-and-authorizations-api/#authenticating-using-query-parameters

File: src/AbstractCas.php (hunk 2 above)
Comment (drupol) [reply src/AbstractCas.php]: I would then prefer to do:

```php
    public function supportAuthentication(array $parameters = []): bool
    {
        return array_key_exists('ticket', $parameters) || Uri::hasParams($this->getServerRequest()->getUri(), 'ticket');
    }

```

File: src/AbstractCas.php (hunk 1 above)
Comment (drupol) [reply src/AbstractCas.php:155]: I guess this will be fixed with the other comment in this PR.

Now, regarding your link and as far as I understand.
I understand Github and why they are doing that, but there is a difference with the CAS protocol. The token which is sent in `ticket` as a parameter in the URL is a one time token. I think there's a difference, security wise, between a one-time token and a permanent access token.

**Comment by drupol** [comment]
This is much better :-)

File: src/CasInterface.php [hunk 4]
```diff
@@ -16,10 +16,13 @@ interface CasInterface
     /**
      * Authenticate the request.
      *
+     * @param array[] $parameters
```
Comment (drupol) [inline src/CasInterface.php]: The type should be `string[]` I guess ?

**Comment by drupol** [comment]
Thanks @samsonradu !!!
